### DISCUSSION TITLE
Tiles on a given A4 sheet

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#7858] [Digital Image Processing]
Centos 8

Version: ImageMagick 7.1.1-6 Q16-HDRI x86_64 21020 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): bzlib djvu fontconfig freetype jbig jng jp2 jpeg lzma pangocairo png tiff webp x xml zip zlib
Compiler: gcc (8.5)

I have a A4 size background image. I want to fill it with 6x7 tiles of individual images with 10px gap between rows and cols. If possible, I want to reduce the opacity of the background image before putting the tiles.

Apart from this, I also want to write the name of the image at the bottom of the tile, without the image extension.

Thanks,

**Comment by fmw42** [comment]
How big are your 6x7 tiles in pixels?  How many are there that you want to put into the A4 image?  Do you need to use all of them and if so can they be resized? If no resizing, then you might not be able to fit them all so they would need to be cropped.

One suggestion would be to first put the filename labels on each image. Then use magick montage to make one large image of all your tiles with your 10 px gaps. Then either crop or resize so that you then fill your 6x7 size image. Then modify the A4 background image to reduce its opacity. Then finally composite the montage created image over your A4 background image.

Can you post your A4 image and the image tiles you want to use? You can zip them all and post the zip file.

**Reply by srini8080** [reply]
Fred, I thought I could do it all in one single line. Tiles can be resized. I have 10,000 images.

![mothergoosevillagecover](https://github.com/user-attachments/assets/6e40eb9a-22cc-42d5-b88e-c43d0b385b2d)
This is the image for A4 background. Can be resized.

![butterlifes-1](https://github.com/user-attachments/assets/c6b70bd1-596c-4012-84b5-e507e84f8569)
This is the sample tile

I am looking to put all the tiles into a pdf file, with each page having 6x7 tiles (6x7 can be changed to anything)

Thanks,

**Reply by fmw42** [reply]
Is 6x7 the aspect ratio or the number of tiles, 6 horizontal and 7 vertical, to be combined for any given page?  Will the same background image be used for each page?

I doubt it can done in one IM command line

You cannot resize the A4 and keep it of A4 pixel dimensions. So the tiles would need to be resized.

**Reply by fmw42** [reply]
Your A4 images is 1024x1024 pixels. But an A4 image must be 595x842.  1024x1024 is square, but 595x842 is rectangular.  So you would need to scale and crop or distort the background image if you want to keep it A4 proper dimensions.

Please clarify what you want to do.

**Comment by srini8080** [comment]
As you can see there are 2 components. A4 size background image and then images for tiles.

We can distort the background to A4 size. A4 is important.

Tiles should maintain aspect ratio of A4. So 1024x1024 images can be resized to maintain A4 aspect ratio.

So I want the tiles to be resized and put on the background image with opacity. PDF output. So for 10,000 images, each page 6x7, roughly 239 pages.

The 6x7 is flexible. Can be changed depending on image resizing,

Thanks

**Reply by fmw42** [reply]
Resizing the tiles and background to A4 dimensions would distort the images badly.  You would have to crop the images first to A4 aspect ratio, then resize to A4 dimensions.

Is 6x7 the aspect ratio or the number of tiles horizontally x vertically?

**Comment by srini8080** [comment]
---> Resizing the tiles and background to A4 dimensions would distort the images badly. You would have to crop the images first to A4 aspect ratio, then resize to A4 dimensions.

Got it, that should be fine. Can create images that are A4.

6 rows, 7 columns.

Thanks,

**Reply by fmw42** [reply]
Either each image needs to be cropped or they need to be combined into 6x7 grid and then cropped and resized. In the former, they would not match at the edges. If combined first, then the pages would not match at the edges.

I would suggest the latter. But note you want a 10 px spacing, which means they would not form a continuous image.  Also if combining does not have the same aspect ratio as an A4, then it would need to be cropped also.

What do you want to do? 

Also how are the images named?  How do you know which images should form each 6x7 combination for any given A4 result?

**Reply by srini8080** [reply]
As you suggested, later is better,

Images are named 1.jpg, 2.jpg etc. So I want the label as "1", "2" etc

Thanks

**Reply by fmw42** [reply]
Are they numbered 1, 2, ... 42 for each 6x7 page? Or are they numbered 1,2, ... 1000 and one has to take them 67 at a time until reaching 1000?

**Reply by fmw42** [reply]
With the latter approach do you want the combined page to be resized and cropped to the A4 size or do you want it resized and padded so that all 42 images show fully in the A4 result?

What color do you want the gaps to be? Or should it be transparent to see the background A4 image?

Are your 6x7 images transparent? If not, why do you need the A4 background image?

You really need to think your process through more fully with the details rather than me having to keep asking more questions!

**Reply by srini8080** [reply]
numbered 1....10000 and should take 42 per page until done.

Resized and padded so that all 42 images show fully in the A4 result. The gaps need not have color as the background is a image. It is there so that the gaps will show a background.

Sorry about the less details

Thanks

**Reply by fmw42** [reply]
Please provide your background image and your first 84 images. Zip them all into one file and post here. Or post to Google or some other hosting platform such as dropbox, etc and post the link.

**Reply by fmw42** [reply]
Just to confirm, you want each of the 6x7 image names with or without suffix shown centered at the bottom of its image and in the image rather than appended below. Is that correct?  Is there a font you want to use and font pointsize?

**Reply by fmw42** [reply]
Please answer my last two questions

**Comment by srini8080** [comment]
Correct. The default font and pointsize 14.

Thanks.

**Reply by fmw42** [reply]
The default font on each system may be different.  Please specify a font

**Comment by srini8080** [comment]
Arial is default on mine. Arial is fine

Thanks for your patience.

**Reply by fmw42** [reply]
What color do you want for the text?

**Comment by srini8080** [comment]
Black. ;-)

**Reply by fmw42** [reply]
What about my request for the first 84 images and the background image?

**Comment by srini8080** [comment]
Arial Font and Text in Black with pointsize 15 for the label/caption, aligned center. The text is the name of the file without extension.

Thanks,

**Reply by fmw42** [reply]
What about my request for the first 84 images and the background image?

**Comment by srini8080** [comment]
[84images.zip](https://github.com/user-attachments/files/18265460/84images.zip)

I had to reduce the depth of the images to make them all compress under 25MB.

Thanks

**Reply by fmw42** [reply]
Please rename your images without spaces.  Also you said they were numbered image-1, image2 … image-84 etc.  Your images do not seem to be numbered is sequence. Please number them in sequential order. Or do you just want them taken in sequence according to how they are listed in the folder, i.e. alphabetic order.

**Reply by fmw42** [reply]
You also did not supply the corresponding background A4 image.

**Reply by fmw42** [reply]
If you are going to renumber them, then please use leading zeros so that they end up in alphabetic order. For example if you have 1000 images, then

`image-0001.png image-0002.png .... image-0010.png .... image-0100.png .... image-1000.png`

**Comment by srini8080** [comment]
Here is the updated zip file with images and also the background 2480x3508 A4 size
[84Images.zip](https://github.com/user-attachments/files/18268523/84Images.zip)

Thanks,

**Reply by fmw42** [reply]
Your background image at 2480 × 3508 has the correct aspect ratio, but is not an A4 size.  A4 is 595x842 at 72 dpi.  It is also distorted. Furthermore, your images are 1024x1024 dimensions.

I presumed that you wanted to keep A4 size at 595x842 and not to distort, but to resize and crop so that it is not distorted and has the proper size at 72 dpi.

Please clarify!  Making each page of the pdf the size of your background image will make a huge PDF file and possibly too large for RAM size during processing.  

1000*2480*3508*3 (3 color channels) is 26,099,520,000 = 26 GBytes

**Reply by srini8080** [reply]
Distorted is fine as the images will cover most of the background. 

maybe a bash script taking 42 at a time (6x7) and  appending the label to each image, then converting into pdf and then uniting all pdfs?

Not sure. You are the best judge. I have 32GB RAM

**Reply by fmw42** [reply]
So please confirm that you want the final pdf pages to each be 2480 × 3508 and that you want all of the 1024x1024 images to be expanded to that size, which will blur the images by the enlargement.  

Would it not be better to have the background resized from your original undistorted image to an A4 aspect ratio closer to 1024x1024 in dimensions?

**Reply by fmw42** [reply]
Saving the pdfs and combining at the end does not help as the final pdf needs all pdfs in memory, if you do not have enough RAM and do not use disk processing which would slow it down dramatically.

**Reply by fmw42** [reply]
I would suggest this size for the background as your original background was 1024x1024 and all your images are 1024x1024.  This way the aspect is still A4, the size of the new background is closer to the input (less blurring) and there is no distortion (just cropping) and processing will be faster and resulting PDF will be smaller file size.

![background](https://github.com/user-attachments/assets/7e9f2eff-7d2e-40ac-a108-62f27d4079fc)

**Reply by srini8080** [reply]
perfect, you are the best.

Uniting pdfs with pdfunite of 1000 pages took 34 seconds.

**Reply by fmw42** [reply]
A custom tool for processing PDFs is going to be much faster than scripting Imagemagick with multiple commands needed to do the processing. I won't be able to do any testing until Tue/Wed.

**Comment by srini8080** [comment]
No hurry. Thanks for your help.

**Comment by fmw42** [comment]
Does this do what you want?  I used the images in the 84Images folder and your original 1024x1024 background image that you posted near the top of this topic. 

I use pipes between IM commands so that intermediate files are not created. I loop over each input image. I then loop over 6x7=42 images at a time. Then I first label all the input images using magick and pipe that to magick montage where I create each page of 6x7 image. Then I pipe that to magick where I resize and crop the background image to the output width and height and also resize the output to the desired output width and height and save the resulting pages as PDF output.

```
# create a new folder for the output.
# put the original background image in the output folder or
# any other folder but the input folder.
# change the 3 paths below.
# the output will be named for the input images without the hyphens, numbers and 
# and suffix using the first input file name.
path2inputfolder="/Users/fred/desktop/84Images"
path2outputfolder="/Users/fred/desktop/testing2"
path2backgroundimage="/Users/fred/desktop/testing2/background.jpeg"
width=1024   #output width
height=1449  #output height
tile_horizontal=6
tile_vertical=7
cd
cd $path2inputfolder
items=(`ls`)
echo "${items[@]}"
echo ""
filename=`echo ${items[0]} | cut -d- -f1`
group_size=$((tile_horizontal * tile_vertical))
echo $filename
echo $group_size
echo ""
( for ((i = 0; i < ${#items[@]}; i += group_size)); do
group=""
for ((j = 0; j < group_size; j++)); do
group=`echo "$group ${items[i + j]}"`
done
echo >&2 "$group"
echo >&2 ""
magick $group -fill black -font Arial -pointsize 14 -gravity south -annotate +0+0 "%t" miff:- |\
magick montage - -background none -geometry +10+10 -tile "${tile_horizontal}x${tile_vertical}" miff:- |\
magick \( $path2backgroundimage -resize ${height}x${height} -background none -gravity center -extent ${width}x${height} +repage \) \
\( - -resize ${width}x${height} -background none -gravity center -extent ${width}x${height} +repage \) \
-gravity center -compose over -composite miff:- 
done ) | magick - $path2outputfolder/$filename.pdf

```

<br>

[image.pdf.zip](https://github.com/user-attachments/files/18276476/image.pdf.zip)

**Reply by srini8080** [reply]
great! thanks much.

Just to make it more aesthetic, if I want to put each image with rounded rectangle, what do I do?

**Comment by fmw42** [comment]
It will be slower. But try this

```
# create a new folder for the output.
# change the 3 paths below.
# the output will be named for the input images without the hyphens,  
# numbers and suffix using the first input file name containing a hyphen.
path2inputfolder="/Users/fred/desktop/84Images"
path2outputfolder="/Users/fred/desktop/testing2"
path2backgroundimage="/Users/fred/desktop/testing2/background.jpeg"
width=1024   #output width
height=1449  #output height
tile_horizontal=6
tile_vertical=7
roundamt=100
path2tempfolder="${path2inputfolder%/*}/tmpdir"
echo $path2tempfolder
mkdir "${path2tempfolder}"
if [ ! -d "$path2tempfolder" ]; then
  echo "$path2tempfolder does not exist"
else
cp -rf $path2inputfolder/* $path2tempfolder
fi
cd
cd $path2tempfolder
items=(`grep -r "-" * | cut -d\  -f3`)
echo "${items[@]}"
echo ""
filename=`echo ${items[@]} | tr " " "\n" | grep -m 1 "-"  | cut -d- -f1`
group_size=$((tile_horizontal * tile_vertical))
echo $filename
echo $group_size
echo ""
( for ((i = 0; i < ${#items[@]}; i += group_size)); do
group=""
for ((j = 0; j < group_size; j++)); do
magick "${items[i + j]}" \
\( +clone -channel a -separate +channel -negate \
\( -size ${roundamt}x${roundamt} xc:black \
-draw "fill white circle ${roundamt},${roundamt} ${roundamt},0" -write mpr:arc +delete \) \
\( mpr:arc \) -gravity northwest -composite \
\( mpr:arc -flip \) -gravity southwest -composite \
\( mpr:arc -flop \) -gravity northeast -composite \
\( mpr:arc -rotate 180 \) -gravity southeast -composite \) \
-alpha off -compose CopyOpacity -composite "${items[i + j]}"
group=`echo "$group ${items[i + j]}"`
done
echo >&2 "$group"
echo >&2 ""
magick $group -fill black -font Arial -pointsize 14 -gravity south -annotate +0+0 "%t" miff:- |\
magick montage - -background none -geometry +10+10 -tile "${tile_horizontal}x${tile_vertical}" miff:- |\
magick \( $path2backgroundimage -resize ${height}x${height} -background none -gravity center -extent ${width}x${height} +repage \) \
\( - -resize ${width}x${height} -background none -gravity center -extent ${width}x${height} +repage \) \
-gravity center -compose over -composite miff:- 
done ) | magick - $path2outputfolder/$filename.pdf
rm -rd $path2tempfolder

```

<br>

[image.pdf.zip](https://github.com/user-attachments/files/18279117/image.pdf.zip)

**Comment by srini8080** [comment]
works like a charm

Thanks much
